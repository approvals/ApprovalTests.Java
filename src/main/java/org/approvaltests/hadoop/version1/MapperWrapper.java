package org.approvaltests.hadoop.version1;

import java.io.IOException;

import org.apache.hadoop.mapreduce.Mapper;

public class MapperWrapper<KeyIn, ValueIn, KeyOut, ValueOut> extends SmartMapper<KeyIn, ValueIn, KeyOut, ValueOut>
{
  private final Class<KeyIn>                             keyIn;
  private final Class<ValueIn>                           valueIn;
  private final Class<KeyOut>                            keyOut;
  private final Class<ValueOut>                          valueOut;
  private final Mapper<KeyIn, ValueIn, KeyOut, ValueOut> mapper;
  public MapperWrapper(Mapper<KeyIn, ValueIn, KeyOut, ValueOut> mapper, Class<KeyIn> keyIn,
      Class<ValueIn> valueIn, Class<KeyOut> keyOut, Class<ValueOut> valueOut)
  {
    this.mapper = mapper;
    this.keyIn = keyIn;
    this.valueIn = valueIn;
    this.keyOut = keyOut;
    this.valueOut = valueOut;
  }
  @Override
  public Class<KeyIn> getKeyInType()
  {
    return keyIn;
  }
  @Override
  public Class<ValueIn> getValueInType()
  {
    return valueIn;
  }
  @Override
  public Class<KeyOut> getKeyOutType()
  {
    return keyOut;
  }
  @Override
  public Class<ValueOut> getValueOutType()
  {
    return valueOut;
  }
  @Override
  public void run(org.apache.hadoop.mapreduce.Mapper.Context context) throws IOException, InterruptedException
  {
    mapper.run(context);
  }
}
