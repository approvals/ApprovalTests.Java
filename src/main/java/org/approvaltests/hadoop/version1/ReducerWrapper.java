package org.approvaltests.hadoop.version1;

import java.io.IOException;

import org.apache.hadoop.mapreduce.Reducer;

public class ReducerWrapper<KeyIn, ValueIn, KeyOut, ValueOut>
    extends
      SmartReducer<KeyIn, ValueIn, KeyOut, ValueOut>
{
  private final Class<KeyIn>                              keyIn;
  private final Class<ValueIn>                            valueIn;
  private final Class<KeyOut>                             keyOut;
  private final Class<ValueOut>                           valueOut;
  private final Reducer<KeyIn, ValueIn, KeyOut, ValueOut> reducer;
  public ReducerWrapper(Reducer<KeyIn, ValueIn, KeyOut, ValueOut> reducer, Class<KeyIn> keyIn,
      Class<ValueIn> valueIn, Class<KeyOut> keyOut, Class<ValueOut> valueOut)
  {
    this.reducer = reducer;
    this.keyIn = keyIn;
    this.valueIn = valueIn;
    this.keyOut = keyOut;
    this.valueOut = valueOut;
  }
  @Override
  public Class<KeyIn> getKeyInType()
  {
    return keyIn;
  }
  @Override
  public Class<ValueIn> getValueInType()
  {
    return valueIn;
  }
  @Override
  public Class<KeyOut> getKeyOutType()
  {
    return keyOut;
  }
  @Override
  public Class<ValueOut> getValueOutType()
  {
    return valueOut;
  }
  @Override
  public void run(Context context) throws IOException, InterruptedException
  {
    reducer.run(context);
  }
}
