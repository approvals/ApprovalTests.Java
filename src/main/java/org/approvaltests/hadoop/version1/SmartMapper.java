package org.approvaltests.hadoop.version1;

import org.apache.hadoop.mapreduce.Mapper;
import org.approvaltests.hadoop.Transform;

public abstract class SmartMapper<KeyIn, ValueIn, KeyOut, ValueOut>
    extends
      Mapper<KeyIn, ValueIn, KeyOut, ValueOut> implements Transform<KeyIn, ValueIn, KeyOut, ValueOut>
{
}
