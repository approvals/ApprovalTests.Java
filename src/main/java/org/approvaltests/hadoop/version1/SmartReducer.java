package org.approvaltests.hadoop.version1;

import org.apache.hadoop.mapreduce.Reducer;
import org.approvaltests.hadoop.Transform;

public abstract class SmartReducer<KeyIn, ValueIn, KeyOut, ValueOut>
    extends
      Reducer<KeyIn, ValueIn, KeyOut, ValueOut> implements Transform<KeyIn, ValueIn, KeyOut, ValueOut>
{
}