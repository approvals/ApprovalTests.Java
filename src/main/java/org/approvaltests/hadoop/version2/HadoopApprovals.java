package org.approvaltests.hadoop.version2;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import org.apache.hadoop.io.LongWritable;
import org.apache.hadoop.io.Text;
import org.apache.hadoop.mrunit.MapDriver;
import org.apache.hadoop.mrunit.MapReduceDriver;
import org.apache.hadoop.mrunit.ReduceDriver;
import org.approvaltests.Approvals;
import org.approvaltests.hadoop.Echo;
import org.approvaltests.hadoop.PairComparer;
import org.approvaltests.hadoop.WritableUtils;

import com.spun.util.ArrayUtils;

/**
 * To use this package you will need:
 * <ul>
 * <li>hadoop-client-1.0.3.jar</li>
 * <li>hadoop-core-0.20.2.jar</li>
 * <li>hadoop-mapred-test-0.22.0.jar</li>
 * <li>log4j-1.2.15.jar</li>
 * <li>mockito-all-1.8.5.jar</li>
 * <li>mrunit-0.9.0-incubating-hadoop1.jar</li>
 * </ul>
 * in your build path.
 * 
 * You will also want to either extended the SmartMapper & SmartWrapper classes,or use the MapperWrapper & ReducerWrapper classes.
 * 
 * @see SmartMapper
 * @see SmartReducer
 * @see MapperWrapper
 * @see ReducerWrapper
 */
@SuppressWarnings({"rawtypes", "unchecked"})
public class HadoopApprovals
{
  public static void verifyMapping(SmartMapper mapper, Object key, Object input) throws Exception
  {
    MapDriver mapDriver = new MapDriver();
    mapDriver.setMapper(mapper);
    Object writableKey = WritableUtils.createWritable(key, mapper.getKeyInType());
    Object writableValue = WritableUtils.createWritable(input, mapper.getValueInType());
    mapDriver.withInput(writableKey, writableValue);
    List results = mapDriver.run();
    Collections.sort(results, PairComparer.INSTANCE);
    String header = String.format("[%s]\r\n\r\n -> maps via %s to -> \r\n", input, mapper.getClass()
        .getSimpleName());
    Approvals.verifyAll(header, results, Echo.INSTANCE);
  }
  public static <KeyIn, ValueIn, KeyOut, ValueOut> void verifyReducer(
      SmartReducer<KeyIn, ValueIn, KeyOut, ValueOut> reducer, Object key, Object... values) throws Exception
  {
    List<ValueIn> list = new ArrayList<ValueIn>();
    for (Object value : values)
    {
      list.add(WritableUtils.createWritable(value, reducer.getValueInType()));
    }
    ReduceDriver reduceDriver = new ReduceDriver<Text, LongWritable, Text, LongWritable>();
    reduceDriver.withInput(WritableUtils.createWritable(key, reducer.getKeyInType()), list);
    reduceDriver.setReducer(reducer);
    List results = reduceDriver.run();
    Collections.sort(results, PairComparer.INSTANCE);
    String header = String.format("(%s, %s)\r\n\r\n -> reduces via %s to -> \r\n", key, list, reducer.getClass()
        .getSimpleName());
    Approvals.verifyAll(header, results, Echo.INSTANCE);
  }
  public static void verifyMapReduce(SmartMapper mapper, SmartReducer reducer, Object key, Object input)
      throws Exception
  {
    MapDriver mapDriver = new MapDriver();
    mapDriver.setMapper(mapper);
    MapReduceDriver mapReduceDriver = new MapReduceDriver();
    mapReduceDriver.setMapper(mapper);
    Object writableKey = WritableUtils.createWritable(key, mapper.getKeyInType());
    Object writableValue = WritableUtils.createWritable(input, mapper.getValueInType());
    mapDriver.withInput(writableKey, writableValue);
    List results = mapDriver.run();
    Collections.sort(results, PairComparer.INSTANCE);
    mapReduceDriver = new MapReduceDriver<LongWritable, Text, Text, LongWritable, Text, LongWritable>();
    writableKey = WritableUtils.createWritable(key, mapper.getKeyInType());
    writableValue = WritableUtils.createWritable(input, mapper.getValueInType());
    mapReduceDriver.withInput(writableKey, writableValue);
    mapReduceDriver.setMapper(mapper);
    mapReduceDriver.setReducer(reducer);
    List finalResults = mapReduceDriver.run();
    String text = String.format(
        "[%s]\r\n\r\n -> maps via %s to -> \r\n\r\n%s\r\n\r\n -> reduces via %s to -> \r\n\r\n%s", input, mapper
            .getClass().getSimpleName(), ArrayUtils.toString(results, Echo.INSTANCE), reducer.getClass()
            .getSimpleName(), ArrayUtils.toString(finalResults, Echo.INSTANCE));
    Approvals.verify(text);
  }
}
