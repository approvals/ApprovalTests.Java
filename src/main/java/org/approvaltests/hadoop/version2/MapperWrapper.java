package org.approvaltests.hadoop.version2;

import java.io.IOException;

import org.apache.hadoop.mapred.JobConf;
import org.apache.hadoop.mapred.Mapper;
import org.apache.hadoop.mapred.OutputCollector;
import org.apache.hadoop.mapred.Reporter;

public class MapperWrapper<KeyIn, ValueIn, KeyOut, ValueOut>
    implements
      SmartMapper<KeyIn, ValueIn, KeyOut, ValueOut>
{
  private final Class<KeyIn>                             keyIn;
  private final Class<ValueIn>                           valueIn;
  private final Class<KeyOut>                            keyOut;
  private final Class<ValueOut>                          valueOut;
  private final Mapper<KeyIn, ValueIn, KeyOut, ValueOut> mapper;
  public MapperWrapper(Mapper<KeyIn, ValueIn, KeyOut, ValueOut> mapper, Class<KeyIn> keyIn,
      Class<ValueIn> valueIn, Class<KeyOut> keyOut, Class<ValueOut> valueOut)
  {
    this.mapper = mapper;
    this.keyIn = keyIn;
    this.valueIn = valueIn;
    this.keyOut = keyOut;
    this.valueOut = valueOut;
  }
  @Override
  public Class<KeyIn> getKeyInType()
  {
    return keyIn;
  }
  @Override
  public Class<ValueIn> getValueInType()
  {
    return valueIn;
  }
  @Override
  public Class<KeyOut> getKeyOutType()
  {
    return keyOut;
  }
  @Override
  public Class<ValueOut> getValueOutType()
  {
    return valueOut;
  }
  @Override
  public void map(KeyIn arg0, ValueIn arg1, OutputCollector<KeyOut, ValueOut> arg2, Reporter arg3)
      throws IOException
  {
    mapper.map(arg0, arg1, arg2, arg3);
  }
  @Override
  public void configure(JobConf arg0)
  {
    mapper.configure(arg0);
  }
  @Override
  public void close() throws IOException
  {
    mapper.close();
  }
}
