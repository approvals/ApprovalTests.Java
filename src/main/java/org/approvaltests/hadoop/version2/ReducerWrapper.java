package org.approvaltests.hadoop.version2;

import java.io.IOException;
import java.util.Iterator;

import org.apache.hadoop.mapred.JobConf;
import org.apache.hadoop.mapred.OutputCollector;
import org.apache.hadoop.mapred.Reducer;
import org.apache.hadoop.mapred.Reporter;

public class ReducerWrapper<KeyIn, ValueIn, KeyOut, ValueOut>
    implements
      SmartReducer<KeyIn, ValueIn, KeyOut, ValueOut>
{
  private final Class<KeyIn>                              keyIn;
  private final Class<ValueIn>                            valueIn;
  private final Class<KeyOut>                             keyOut;
  private final Class<ValueOut>                           valueOut;
  private final Reducer<KeyIn, ValueIn, KeyOut, ValueOut> reducer;
  public ReducerWrapper(Reducer<KeyIn, ValueIn, KeyOut, ValueOut> reducer, Class<KeyIn> keyIn,
      Class<ValueIn> valueIn, Class<KeyOut> keyOut, Class<ValueOut> valueOut)
  {
    this.reducer = reducer;
    this.keyIn = keyIn;
    this.valueIn = valueIn;
    this.keyOut = keyOut;
    this.valueOut = valueOut;
  }
  @Override
  public Class<KeyIn> getKeyInType()
  {
    return keyIn;
  }
  @Override
  public Class<ValueIn> getValueInType()
  {
    return valueIn;
  }
  @Override
  public Class<KeyOut> getKeyOutType()
  {
    return keyOut;
  }
  @Override
  public Class<ValueOut> getValueOutType()
  {
    return valueOut;
  }
  @Override
  public void reduce(KeyIn arg0, Iterator<ValueIn> arg1, OutputCollector<KeyOut, ValueOut> arg2, Reporter arg3)
      throws IOException
  {
    reducer.reduce(arg0, arg1, arg2, arg3);
  }
  @Override
  public void configure(JobConf arg0)
  {
    reducer.configure(arg0);
  }
  @Override
  public void close() throws IOException
  {
    reducer.close();
  }
}
