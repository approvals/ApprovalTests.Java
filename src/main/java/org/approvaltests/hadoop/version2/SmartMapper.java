package org.approvaltests.hadoop.version2;

import org.apache.hadoop.mapred.Mapper;
import org.approvaltests.hadoop.Transform;

public interface SmartMapper<KeyIn, ValueIn, KeyOut, ValueOut>
    extends
      Mapper<KeyIn, ValueIn, KeyOut, ValueOut>,
      Transform<KeyIn, ValueIn, KeyOut, ValueOut>
{
}
