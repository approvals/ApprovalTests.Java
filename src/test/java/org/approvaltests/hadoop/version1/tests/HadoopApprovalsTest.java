package org.approvaltests.hadoop.version1.tests;

import java.io.IOException;

import junit.framework.TestCase;

import org.apache.hadoop.io.LongWritable;
import org.apache.hadoop.io.Text;
import org.apache.hadoop.mapreduce.Mapper;
import org.apache.hadoop.mapreduce.Reducer;
import org.approvaltests.hadoop.version1.HadoopApprovals;
import org.approvaltests.hadoop.version1.MapperWrapper;
import org.approvaltests.hadoop.version1.ReducerWrapper;
import org.approvaltests.hadoop.version1.SmartMapper;
import org.approvaltests.hadoop.version1.SmartReducer;
import org.approvaltests.reporters.DiffReporter;
import org.approvaltests.reporters.UseReporter;

@UseReporter({DiffReporter.class})
public class HadoopApprovalsTest extends TestCase
{
  public void testMap() throws Exception
  {
    SmartMapper mapper = new MapperWrapper(new WordCountMap(), LongWritable.class, Text.class, Text.class,
        LongWritable.class);
    HadoopApprovals.verifyMapping(mapper, 0, "llew gen  llew");
  }
  public void testReduce() throws Exception
  {
    SmartReducer reducer = new ReducerWrapper(new WordCountReduce(), Text.class, LongWritable.class, Text.class,
        LongWritable.class);
    HadoopApprovals.verifyReducer(reducer, "life", 7, 35);
  }
  public void testMapReducer() throws Exception
  {
    SmartMapper mapper = new MapperWrapper(new WordCountMap(), LongWritable.class, Text.class, Text.class,
        LongWritable.class);
    SmartReducer reducer = new ReducerWrapper(new WordCountReduce(), Text.class, LongWritable.class, Text.class,
        LongWritable.class);
    HadoopApprovals.verifyMapReduce(mapper, reducer, 0, "one fish two fish red fish blue fish");
  }
  private static class WordCountReduce extends Reducer<Text, LongWritable, Text, LongWritable>
  {
    @Override
    protected void reduce(Text token, Iterable<LongWritable> counts, Context context) throws IOException,
        InterruptedException
    {
      long n = 0;
      for (LongWritable count : counts)
        n += count.get();
      context.write(token, new LongWritable(n));
    }
  }
  private static class WordCountMap extends Mapper<LongWritable, Text, Text, LongWritable>
  {
    protected void map(LongWritable offset, Text text, Context context) throws IOException, InterruptedException
    {
      for (String token : text.toString().split("\\s+"))
      {
        context.write(new Text(token), new LongWritable(1));
      }
    }
  }
}
