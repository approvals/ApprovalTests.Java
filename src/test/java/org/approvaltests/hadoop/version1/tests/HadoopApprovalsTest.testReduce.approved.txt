(life, [7, 35])

 -> reduces via ReducerWrapper to -> 



(life, 42)
